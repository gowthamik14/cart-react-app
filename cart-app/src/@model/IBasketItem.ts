import IProduct from './IProduct';

export default interface IBasketItem extends IProduct {
  quantity: number;
}
