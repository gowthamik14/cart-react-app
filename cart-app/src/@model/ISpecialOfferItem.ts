export default interface ISpecialOfferItem {
  id: number;
  name: string;
  savingAmount: number;
}
