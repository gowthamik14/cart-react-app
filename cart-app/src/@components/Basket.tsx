import IBasketItem from '@model/IBasketItem';
import ISpecialOfferItem from '@model/ISpecialOfferItem';
import React from 'react';
import BasketItem from './BasketItem';
import { H4 } from './lib/typography';
import { getSavingsAmount, getTotalCost } from '@utils';

const Basket = ({
  products,
  subTotal,
  specialOffers,
  increaseQuantity,
  decreaseQuantity,
}: {
  products: IBasketItem[];
  subTotal: number;
  specialOffers: ISpecialOfferItem[];
  increaseQuantity: (basketItem: IBasketItem) => void;
  decreaseQuantity: (basketItem: IBasketItem) => void;
}): JSX.Element => {
  return (
    <div className="p-2">
      <H4>Basket</H4>
      <hr />
      <ul>
        {products.map((product) => (
          <BasketItem
            key={product.id}
            basketItem={product}
            increaseQuantity={increaseQuantity}
            decreaseQuantity={decreaseQuantity}
            savingAmount={
              specialOffers.find((i) => i.id === product.id)?.savingAmount ?? 0
            }
          />
        ))}
      </ul>
      <div className="flex justify-between items-center">
        <span className="text-1xl">Sub Total: </span>
        <span className="ml-4 p-2 w-15 text-right text-blue rounded-md">
          {`£ ${(subTotal / 100).toFixed(2)}`}
        </span>
      </div>
      <div className="flex justify-between items-center">
        <span className="text-1xl">Savings: </span>
        <span className="ml-4 p-2 w-15 text-right text-blue rounded-md">
          {`£ ${(getSavingsAmount(specialOffers) / 100).toFixed(2)}`}
        </span>
      </div>

      <div className="flex justify-between items-center">
        <span className="text-1xl">Total Amount: </span>
        <span className="ml-4 p-2 w-15 text-right text-blue rounded-md">
          {`£ ${(getTotalCost(specialOffers, subTotal) / 100).toFixed(2)}`}
        </span>
      </div>
    </div>
  );
};

export default Basket;
