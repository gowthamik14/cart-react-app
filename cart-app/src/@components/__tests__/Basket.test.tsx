import React from 'react';
import { render } from '@testing-library/react';
import Basket from '@components/Basket';
import ISpecialOfferItem from '@model/ISpecialOfferItem';
import IBasketItem from '@model/IBasketItem';

test('renders learn react link', () => {
  const products: IBasketItem[] = [
    {
      name: 'cheese',
      price: 20,
      quantity: 3,
      id: 1,
    },
    {
      name: 'Butter',
      price: 20,
      quantity: 3,
      id: 2,
    },
  ];

  const specialOffers: ISpecialOfferItem[] = [
    {
      name: 'cheese',
      savingAmount: 10,

      id: 1,
    },
    {
      name: 'Butter',
      savingAmount: 30,
      id: 2,
    },
  ];

  const subTotal: number = 60;
  const increaseQuantity = jest.fn();
  const decreaseQuantity = jest.fn();
  const { container } = render(
    <Basket
      products={products}
      specialOffers={specialOffers}
      subTotal={subTotal}
      increaseQuantity={increaseQuantity}
      decreaseQuantity={decreaseQuantity}
    />
  );

  expect(container).toMatchSnapshot();
  
});
