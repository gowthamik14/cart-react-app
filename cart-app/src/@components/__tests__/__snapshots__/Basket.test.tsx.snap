// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders learn react link 1`] = `
<div>
  <div
    class="p-2"
  >
    <span
      class="text-3xl"
    >
      Basket
    </span>
    <hr />
    <ul>
      <li
        class="flex flex-col list-decimal list-inside border-b-2 border-gray-100	p-2"
      >
        <div
          class="flex justify-between items-center"
        >
          <span
            class="text-lg"
          >
            cheese
          </span>
          <div>
            <span
              class="text-gray-400 pr-1"
            >
              £
            </span>
            <span
              class="text-lg"
            >
              0.20
            </span>
          </div>
          <span>
            <button
              class="ml-4 p-2 bg-blue-400 text-white rounded-md focus:outline-none focus:ring-2 ring-blue-200 ring-offset-2"
            >
              +
            </button>
            <span
              class="ml-4 p-2 w-10 text-center text-blue rounded-md"
            >
              3
            </span>
            <button
              class="ml-4 p-2  bg-white-400 text-blue rounded-md border-2 border-blue-400 focus:outline-blue focus:ring-2 ring-blue-200 ring-offset-2"
            >
              -
            </button>
          </span>
        </div>
        <div>
          <div
            class="flex flex-col justify-center"
          >
            <div
              class="flex justify-end my-2 border-b-2 border-gray-100 text-gray-500"
            >
              <div
                class="mx-2"
              >
                Item price 
              </div>
              <div>
                £0.20 * 3 = £0.60
              </div>
            </div>
            <div
              class="flex justify-end items-center my-2 border-b-2 border-gray-100 text-red-500"
            >
              <div
                class="mx-2"
              >
                Savings
              </div>
              <div>
                £0.10
              </div>
            </div>
            <div
              class="flex justify-end my-2 items-center"
            >
              <div
                class="mx-2"
              >
                Item cost 
              </div>
              <div>
                £0.50
              </div>
            </div>
          </div>
        </div>
      </li>
      <li
        class="flex flex-col list-decimal list-inside border-b-2 border-gray-100	p-2"
      >
        <div
          class="flex justify-between items-center"
        >
          <span
            class="text-lg"
          >
            Butter
          </span>
          <div>
            <span
              class="text-gray-400 pr-1"
            >
              £
            </span>
            <span
              class="text-lg"
            >
              0.20
            </span>
          </div>
          <span>
            <button
              class="ml-4 p-2 bg-blue-400 text-white rounded-md focus:outline-none focus:ring-2 ring-blue-200 ring-offset-2"
            >
              +
            </button>
            <span
              class="ml-4 p-2 w-10 text-center text-blue rounded-md"
            >
              3
            </span>
            <button
              class="ml-4 p-2  bg-white-400 text-blue rounded-md border-2 border-blue-400 focus:outline-blue focus:ring-2 ring-blue-200 ring-offset-2"
            >
              -
            </button>
          </span>
        </div>
        <div>
          <div
            class="flex flex-col justify-center"
          >
            <div
              class="flex justify-end my-2 border-b-2 border-gray-100 text-gray-500"
            >
              <div
                class="mx-2"
              >
                Item price 
              </div>
              <div>
                £0.20 * 3 = £0.60
              </div>
            </div>
            <div
              class="flex justify-end items-center my-2 border-b-2 border-gray-100 text-red-500"
            >
              <div
                class="mx-2"
              >
                Savings
              </div>
              <div>
                £0.30
              </div>
            </div>
            <div
              class="flex justify-end my-2 items-center"
            >
              <div
                class="mx-2"
              >
                Item cost 
              </div>
              <div>
                £0.30
              </div>
            </div>
          </div>
        </div>
      </li>
    </ul>
    <div
      class="flex justify-between items-center"
    >
      <span
        class="text-1xl"
      >
        Sub Total: 
      </span>
      <span
        class="ml-4 p-2 w-15 text-right text-blue rounded-md"
      >
        £ 0.60
      </span>
    </div>
    <div
      class="flex justify-between items-center"
    >
      <span
        class="text-1xl"
      >
        Savings: 
      </span>
      <span
        class="ml-4 p-2 w-15 text-right text-blue rounded-md"
      >
        £ 0.40
      </span>
    </div>
    <div
      class="flex justify-between items-center"
    >
      <span
        class="text-1xl"
      >
        Total Amount: 
      </span>
      <span
        class="ml-4 p-2 w-15 text-right text-blue rounded-md"
      >
        £ 0.20
      </span>
    </div>
  </div>
</div>
`;
